feat(backend): Support pipeline loops via tekton custom task

fenglixa opened the pull request:
**Which issue is resolved by this Pull Request:** 
Support pipeline loops via tekton custom task

**Description of your changes:**
Create contorller for tekton custom task to handle the loop cases
Refer to the readme for the detail steps.

**Environment tested:**
```
fengli@fenglis-mbp >> ~/go/src/github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops >>  support-pipeline-loops ● >> kc apply -f examples/pipelinespec-with-run-string-value.yaml
pipelineloop.custom.tekton.dev/echoloop created
run.tekton.dev/simplepipelinelooprun created
 fengli@fenglis-mbp >> ~/go/src/github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops >>  support-pipeline-loops ● >> kc get run
NAME                    SUCCEEDED   REASON    STARTTIME   COMPLETIONTIME
simplepipelinelooprun   Unknown     Running   19s
 fengli@fenglis-mbp >> ~/go/src/github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops >>  support-pipeline-loops ● >> kc get pr |grep simplepipelinelooprun
simplepipelinelooprun-00001-whw8h   True        Succeeded            30s         20s
simplepipelinelooprun-00002-499tr   True        Succeeded            20s         11s
simplepipelinelooprun-00003-27lqq   True        Succeeded            11s         2s
```

* Python Version (use `python --version`):
* Tekton Version (use `tkn version`):
* Kubernetes Version (use `kubectl version`):
* OS (e.g. from `/etc/os-release`):

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.

- [x] Do you want this pull request (PR) cherry-picked into the current release branch?
    
    If yes, use one of the following options:
  
    * **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update.
    *  After this PR is merged, create a cherry-pick PR to add these changes to the release branch. (For more information about creating a cherry-pick PR, see the [Kubeflow Pipelines release guide](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#option--git-cherry-pick).)

Tomcli commented on tekton-catalog/pipeline-loops/examples/simplepipelineloop.yaml:
```diff
@@ -0,0 +1,82 @@
+  tasks:
+  - name: simpletask
+    taskRef:
+      name: simpletask
```
Can you use pipeline.taskRef to refer to custom task? Also, does pipeline.taskSpec support custom task?
Since custom task is using the alpha api, it's better to put it under the pipeline.taskSpec if possible to avoid the Tekton client version conflict.

fenglixa replied:
Yes, use pipeline.taskRef is the only case of using custom task.

Tomcli commented on tekton-catalog/pipeline-loops/README.md:
```diff
@@ -0,0 +1,32 @@
+- Edit feature-flags configmap, ensure "data.enable-custom-tasks" is "true":
+`kubectl edit cm feature-flags -n tekton-pipelines`
+
+- Run the E2E cases: `kubectl apply -f examples/dsl-loop-example-basic.yaml`
```
I guess the dsl will come in another PR?

fenglixa replied:
Yes, the dsl will come in another PR, with the issue traced:https://github.com/kubeflow/kfp-tekton/issues/413

fenglixa replied:
Oh, seems I missed the dsl-loop-example-basic.yaml here, let me add it now.

fenglixa replied:
@Tomcli I added loop-example-basic.yaml of the E2E case now, please help check. Thanks

Tomcli reviewed (commented):
thanks @fenglixa. I can run the custom task with "run" crd. However, we should discuss about how we can move this custom task to pipeline run and execute it with the beta Tekton client. 

In my opinion, it would be better if we can change the Tekton source code to include the custom task into the pipeline.taskSpec. Because the current KFP logics for TTL, cache, and other features are based on the tekton beta client.

Tomcli commented on tekton-catalog/pipeline-loops/README.md:
```diff
@@ -0,0 +1,32 @@
+  `go mod vendor; go mod tidy; bash hack/update-codegen.sh`
+- Install&config [KO](https://github.com/google/ko)
+- Install Exception Handler
+  `ko apply -f config/`
```
can you also add a docker image build so we can run it on our ci and for users who have problem with ko?

fenglixa replied:
@ Tomcli I addressed you comments, by adding an dockerfile, makefile, and the readme steps for user to build his own docker file to install the controller

Tomcli commented on tekton-catalog/pipeline-loops/README.md:
```diff
@@ -0,0 +1,32 @@
+
+# Take a try
+- Generate the code of client
+  `go mod vendor; go mod tidy; bash hack/update-codegen.sh`
```
I have to upgrade my local machine go version due to some other project. On go 1.15.3 I'm see this error on my go mod vendor.
```
go: finding module for package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/informers/pipelineloop/v1alpha1/pipelineloop
go: finding module for package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/clientset/versioned
go: finding module for package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/client
go: finding module for package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/client/fake
go: finding module for package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/informers/pipelineloop/v1alpha1/pipelineloop/fake
go: finding module for package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/listers/pipelineloop/v1alpha1
github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun imports
	github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/clientset/versioned: module github.com/kubeflow/kfp-tekton@latest found (v0.5.0), but does not contain package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/clientset/versioned
github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun imports
	github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/client: module github.com/kubeflow/kfp-tekton@latest found (v0.5.0), but does not contain package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/client
github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun imports
	github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/informers/pipelineloop/v1alpha1/pipelineloop: module github.com/kubeflow/kfp-tekton@latest found (v0.5.0), but does not contain package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/informers/pipelineloop/v1alpha1/pipelineloop
github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun imports
	github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/listers/pipelineloop/v1alpha1: module github.com/kubeflow/kfp-tekton@latest found (v0.5.0), but does not contain package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/listers/pipelineloop/v1alpha1
github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun tested by
	github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun.test imports
	github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/client/fake: module github.com/kubeflow/kfp-tekton@latest found (v0.5.0), but does not contain package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/client/fake
github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun tested by
	github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun.test imports
	github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/informers/pipelineloop/v1alpha1/pipelineloop/fake: module github.com/kubeflow/kfp-tekton@latest found (v0.5.0), but does not contain package github.com/kubeflow/kfp-tekton/tekton-catalog/pipeline-loops/pkg/client/injection/informers/pipelineloop/v1alpha1/pipelineloop/fake
```

it's probably just my env. can you also populate the pkg/client as part of this PR? Thanks

fenglixa replied:
I commit the code of client as well, to avoid user generate the client code again.

Tomcli reviewed (commented):
Thanks, the examples work great (I used your docker image this time). For the taskSpec question I also posted to Tekton and see what's their direction. I have few questions on the image build but otherwise it looks good to me.

fenglixa commented:
@Tomcli I addressed your comments of building userself docker images instead of using KO, and also the client code to avoid user re-generate it again. Please check. Thanks.

Tomcli commented on tekton-catalog/pipeline-loops/Makefile:
```diff
@@ -0,0 +1,32 @@
+	CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o=${BIN_DIR}/pipelineloop-webhook ./cmd/webhook
+
+images: build-linux
+	docker build --build-arg bin_name=pipelineloop-controller . -t fenglixa/pipelineloop-controller:$(TAG)
```
small nit: change `fenglixa` to a variable (like the tag), you can default it to something generic like `registryname`

fenglixa replied:
Thanks @Tomcli , I addressed your comments.

fenglixa commented:
Thanks @Tomcli , I addressed your comments, is it OK for you now. Will provide another PR for DSL after this PR merged.

Tomcli commented:
/lgtm
FYI for other reviewers: this is just a custom task controller for tekton loops. The actual kfp integration will come in next PR.
